packages/components/src/FormField/FormField.tsx

```diff
@@ -1,426 +1,216 @@
+  useEffect(() => handleValidation(), [error]);
 
   return (
-    <Wrapper
-      className={wrapperClassNames}
-      style={{ ["--formField-maxLength" as string]: maxLength || max }}
-    >
-      <label className={labelClassNames} htmlFor={identifier}>
-        {placeholder || " "}
-      </label>
-      {fieldElement()}
-      {type === "text" && loading && (
-        <span className={styles.postfix}>
-          <Spinner size="small" />
-        </span>
-      )}
-      {type === "select" && (
-        <span className={styles.postfix}>
-          <Icon name="arrowDown" />
-        </span>
-      )}
+    <FieldWrapper {...props} error={error} hasMiniLabel={hasMiniLabel}>
+      <FormLabel {...props} identifier={identifier} />
+      <Controller
+        control={control}
+        name={controlledName}
+        rules={{ ...validations }}
+        defaultValue={value ?? defaultValue ?? ""}
+        render={({
+          onChange: onControllerChange,
+          onBlur: onControllerBlur,
+          name: controllerName,
+          ...rest
+        }) => {
+          const fieldClasses = classnames(styles.formField, {
+            [styles.select]: type === "select",
+          });
+
+          const fieldProps = {
+            ...rest,
+            id: identifier,
+            className: fieldClasses,
+            name: (props.validations || props.name) && controllerName,
+            disabled: disabled,
+            readOnly: readonly,
+            inputMode: keyboard,
+            onChange: handleChange,
+          };
+
+          const textFieldProps = {
+            ...fieldProps,
+            onBlur: handleBlur,
+            onFocus: handleFocus,
+            onKeyDown: handleKeyDown,
+          };
+
+          return renderField();
+
+          function renderField() {
+            switch (type) {
+              case "select":
+                return (
+                  <>
+                    <select {...fieldProps}>{children}</select>
+                    <span className={styles.postfix}>
+                      <Icon name="arrowDown" />
+                    </span>
+                  </>
+                );
+              case "textarea":
+                return (
+                  <textarea
+                    rows={rows}
+                    ref={inputRef as MutableRefObject<HTMLTextAreaElement>}
+                    {...textFieldProps}
+                  />
+                );
+              default:
+                return (
+                  <>
+                    <input
+                      autoComplete={setAutocomplete(autocomplete)}
+                      type={type}
+                      maxLength={maxLength}
+                      max={max}
+                      min={min}
+                      ref={inputRef as MutableRefObject<HTMLInputElement>}
+                      {...textFieldProps}
+                    />
+                    {loading && <FormSpinner />}
+                  </>
+                );
```

Comment:
Part of me feels like splitting the internals of this render callback off as a separate component would be nice. Thoughts? It does lean on a lot of scoped things from above but it would simplify this a decent amount too.

Reply:
This is something i struggled on.

When I pulled this out into its own component I was having issues with the typing of it. The reason being is that we found some issues with the types around the `Controller` component. 

I was unable to type the `{...rest}` spread when pulling this into its own component. 

I did try to make it a little cleaner by bringing the `Spinner` and `Icon` into their respected return statements and thought that helped a bit.

Reply:
Yea, I can see that. Let's leave it as is for now.